src/main/kotlin/org/rust/lang/core/resolve/RustResolveEngine.kt

```diff
@@ -91,32 +134,29 @@ object RustResolveEngine {
             val pathPart = path.pathPart ?: return
 
             val isPlainPathImport = path.mul == null && path.lbrace == null
+
+            // `use foo::bar as baz;`
             if (isPlainPathImport) {
                 val name = path.alias ?: pathPart
                 if (match(name)) {
-                    if (use in visited) {
+                    if (use in visitedImports) {
                         return
                     }
-                    visited += use
+                    visitedImports += use
 
-                    val item = RustResolveEngine.resolve(pathPart, visited).element  ?: return
+                    val item = resolve(pathPart).element  ?: return
                     return found(item)
                 }
             }
-        }
-
 
-        private fun seek(elem: RustDeclaringElement) {
-            seek(listOf(elem))
         }
 
+        private fun seek(elem: RustDeclaringElement) = seek(listOf(elem))
+
         private fun seek(decls: Collection<RustDeclaringElement>) {
             decls.flatMap { it.getBoundElements() }
-                 .forEach { e ->
-                     if (match(e)) {
-                         return found(e)
-                     }
-                 }
+                 .find{match(it)}
+                 ?.let {found(it)}
```

Comment:
You mean something like

```Kotlin
public inline fun <T> T?.then(f: (T) -> Unit): Unit =
    if(this != null)
        f(this)
    else
        Unit
```

?

So that we can write

```Kotlin
            decls.flatMap { it.getBoundElements() }
                 .find { match(it) }
                 .then { found(it) }
```